server/src/main/java/com/cloud/network/element/VpcVirtualRouterElement.java

```diff
@@ -348,6 +348,9 @@ public Provider getProvider() {
 
         //1st time it runs the domain router of the VM shall be returned
         List<DomainRouterVO> routers = super.getRouters(network, dest);
+        if (routers == null) {
```

Comment:
@ravening 
this might cause some issues. 
the codes from line 371 to line 379 which create and deploy virtual routers will be ignored.

Reply:
@ravening 
I noticed the issue you mentioned in the description. 
could you please look for other ways to fix the issue ?

Reply:
@weizhouapache I tried to deploy a new vpc, with tiers and vm in it but the control never reached the line 356 of this file.

Reply:
@ravening
destroy all routers and then deploy vm ?

Reply:
@weizhouapache worked fine as well

Reply:
@ravening 
Sorry I still did not get why you made this change.
you said that there was a libvirt issue when you added new vpc tier. Was the vpc tier added successfully ? I think the vpc tier should not be added.

Reply:
@weizhouapache vpc tier was already added successfully but when i tried to deploy a vm in that tier, it failed because of libvirt issue and also the vm deployment failed saying "internal server error". In the logs I could see that there NPE at this code location. So I navigated to networks -> select vpc tier -> virtual routers, it was empty. for other tiers in the same vpc, i could see VR

Reply:
@ravening 
Got it. so what do you want to solve, the NPE exception ? can you try to change the code which process the PlugNicAnswer ?

Reply:
@weizhouapache yes, I want to fix the NPE ... let me check the code in PlugNicAnswer

Reply:
@weizhouapache `PlugNicAnswer` may not solve the issue. the NPE happens when the code tries to list the VR's for the network but it gets a null rather than empty list or list with size > 0

Reply:
> 

@ravening Did NPE happen when deploy a vm or list routers ?

Reply:
@weizhouapache NPE happened while deploying vm

Reply:
> 
a bit confused. deploy vm -> plug nic to VR -> libvirt issue -> NPE ?

Reply:
yes..

when i try to deploy a vm, it send plugnic command to agent and that failed. after that we get this NPE

Reply:
@weizhouapache you need any more clarifications?

Reply:
@ravening 
I think the normal way to fix NPE is adding a check like below.
```
 
         //1st time it runs the domain router of the VM shall be returned
         List<DomainRouterVO> routers = super.getRouters(network, dest);
-        if (routers.size() > 0) {
+        if (routers != null && routers.size() > 0) {
             return routers;
         }
 
```

then I got following error when I create the 24th vpc tier
`
Unable to add NIC to VM instance {id: \"1693\", name: \"i-2-1693-VM\", uuid: \"f79f121f-65bb-4212-82b8-99f3be9d3654\", type=\"User\"}: com.cloud.exception.ResourceUnavailableException: Resource [DataCenter:1] is unreachable: Unable to apply network rules as a part of network Ntwk[3521|Guest|11] implement`

What is the error in your testing result ?

both options are good to me.

Reply:
Hi @ravening is this PR ready, or any further changes to be done based on the feedback?

Reply:
@weizhouapache I updated the code